src/index.ts

```diff
@@ -1381,6 +1381,38 @@ class Remarkable implements RemarkableApi {
     return await this.bulkMove(hashes, "trash", refresh);
   }
 
+
+  /** permanent delete many hashes */
+  async bulkPurge(
+    hashes: readonly string[],
+    refresh: boolean = false,
+  ): Promise<SimpleEntry[]> {
+    const [rootHash, generation] = await this.#getRootHash(refresh);
+    // Get the raw text of the root entry
+
+    const lines = rootHash.split("\n");
```

Comment:
the root hash should be just a string hash. Similar to other bulk actions, you should need to get the list of entries at that hash with:
```
const entries = await this.raw.getEntries(rootHash)
```
then to do a bulk purge you just need to remove the hashes from that entry set, and create a new entry with only the remaining ones, a. la. the last part of the bulk modify actions.

This way you can take advantage of the api parsing, rather than doing the raw line manipulation yourself, which is what it looks like you're doing.

This also does the v3 check (or if it doesn't, should, so some more code you can rely on general utilities.

Reply:
just look a few lines up:
```
    const [rootEntry, uploadRoot] = await this.raw.putEntries(
      "root",
      newEntries,
    );
    await uploadRoot;
    await this.#putRootHash(rootEntry.hash, generation);
```
the api may not be super clear, but the `rootEntry` has the new root hash, and `uploadRoot` is a promise that resolves when everything is actually uploaded. By splitting these up it allows more efficient async operations, but since you're only doing one upload, you might as well just await uploadRoot immediately.

Reply:
leaving a comment inline, after you change